src/core/thread/topology.hpp

```diff
@@ -862,6 +872,56 @@ class Child : public Neighbor, public IndirectSender::ChildInfo, public DataPoll
 
 #endif // #if OPENTHREAD_CONFIG_CHILD_SUPERVISION_ENABLE
 
+#if OPENTHREAD_CONFIG_TMF_PROXY_MLR_ENABLE
+    /**
+     * This method returns MLR state of an Ip6 address.
+     *
+     * @param[in] aAddressIndex  The Ip6 address index.
+     *
+     * @returns MLR state of the Ip6 address.
+     *
+     */
+    MlrState GetAddressMlrState(uint16_t aAddressIndex);
+
```

Comment:
Follow up on previous approach:

What we want is to get/set the MLR state on a specific "child IPv6 address" as we iterate through the addresses.

How about the following
- We can consider defining a new class/type `Child::Ip6Address` which represents an IPv6 address on a `Child`
  - It provides `const Ip6::Address &GetAddress()` to get the associated IPv6 address, 
  -  `Child &GetChild()` for the associated `Child`, etc
  - It can also provide `SetMlrState()` and `GetMlrState()` to allow MLR state to be set on a specific address of the child 
     - The implementation of these can call some private methods of `Child` to update the masks but that all of this would be hidden from caller).
- We change iterator such that when we iterate using `IterateIp6Addresses()`, we get a `Child::Ip6Address` instead of ab `Ip6::Address`, i.e.
```
   for (Child::Ip6Address &addr : aChild.IterateIp6Addresses(filter))
   {
        // Can use addr.GetAddress() to get the actual `Ip6::Address &`
        // Can use `addr.GetMlrState()` or `addr.SetMleState()` directly on `addr`
        ...
   }
````

Reply:
This sounds good to me overall. My only concern is that having many different `XxxAddress` could make the code less straightforward to understand?

Reply:
If `IterateIp6Addresses` iterates on `Child::Ip6Address &addr`, we will need to construct a fake `Child::Ip6Address` from MLEID.

Reply:
I tried to use `Ip6Address` directly:
```c++
Ip6Address               mIp6Address[kNumIp6Addresses]; ///< Registered IPv6 addresses
``` 

Turnout it's not workable because `Ip6Address` can not contain any additional information (we don't want to increase the memory consumption of Child). 

Then I realized that `Ip6Address` is essential the `AddressIterator` itself. We can even define `GetMlrState` and `SetMlrState` in `AddressIterator` class and it will work.

That being said, I think `IterateIp6Addresses()` should return `Ip6::Address` to have a straightforward interface. Now I tend to keep the code as it is so as to make `AddressIterator` less complex.

Reply:
We don't want to change the `mIp6Address` it should still be `Ip6::Address mIp6Address`.
The `Child::Ip6Address` type would use indireation, i.e. has a pointer/reference to an `Ip6::Address` (which we get from `mIpAddress[]` or it is from iterator `mMeshLocalAddress`). I agree the `AddressIterator` itself can be made to look like `Child::Ip6Address`.

While I like this approach, my concern with it is that it will increase code size (due to all indirection and new type). So I think we can consider it for future and try it out later after this PR is merged.

I would suggest for this PR, we follow the https://github.com/openthread/openthread/pull/5236/files#r459920104 (from comment above), i.e. pass the `aAddress` reference directly to `Get/SetMlrState()` (instead of iter index).

Reply:
I will resolve this PR and revise according to https://github.com/openthread/openthread/pull/5236#discussion_r459920104